packages/cli/src/evaluation/test-runs.controller.ee.ts

```diff
@@ -45,33 +71,44 @@ export class TestRunsController {
 
 	@Get('/:testDefinitionId/runs/:id')
 	async getOne(req: TestRunsRequest.GetOne) {
-		const { id: testRunId, testDefinitionId } = req.params;
-
 		await this.getTestDefinition(req);
 
-		const testRun = await this.testRunRepository.findOne({
-			where: { id: testRunId, testDefinition: { id: testDefinitionId } },
-		});
-
-		if (!testRun) throw new NotFoundError('Test run not found');
-
-		return testRun;
+		return await this.getTestRun(req);
 	}
 
 	@Delete('/:testDefinitionId/runs/:id')
 	async delete(req: TestRunsRequest.Delete) {
-		const { id: testRunId, testDefinitionId } = req.params;
+		const { id: testRunId } = req.params;
 
+		// Check test definition and test run exist
 		await this.getTestDefinition(req);
-
-		const testRun = await this.testRunRepository.findOne({
-			where: { id: testRunId, testDefinition: { id: testDefinitionId } },
-		});
-
-		if (!testRun) throw new NotFoundError('Test run not found');
+		await this.getTestRun(req);
 
 		await this.testRunRepository.delete({ id: testRunId });
 
 		return { success: true };
 	}
+
+	@Post('/:testDefinitionId/runs/:id/cancel')
+	async cancel(req: TestRunsRequest.Cancel, res: express.Response) {
+		if (this.instanceSettings.isMultiMain) {
+			throw new NotImplementedError('Cancelling test runs is not yet supported in multi-main mode');
+		}
+
+		const { id: testRunId } = req.params;
+
+		// Check test definition and test run exist
+		await this.getTestDefinition(req);
+		const testRun = await this.getTestRun(req);
+
+		// Check the state of the test run
+		if (testRun.status !== 'running' && testRun.status !== 'new') {
+			const message = `The test run "${testRunId}" is not running and cannot be cancelled`;
+			throw new ConflictError(message);
+		}
```

Comment:
This is business logic and should not be in the controller. Let's move this to the service instead. Either the whole check and throwing or implement `canBeCancelled`.

And we probably should do the same check within `cancelTestRun`